go/arrow/compute/expression.go

```diff
@@ -0,0 +1,954 @@
+	// null.
+	IsNullLiteral() bool
+	// IsSatisfiable returns true if this expression could evaluate to true
+	IsSatisfiable() bool
+	// FieldRef returns a pointer to the underlying field reference, or nil if
+	// this expression is not a field reference.
+	FieldRef() *FieldRef
+	// Descr returns the shape of this expression will evaluate to including the type
+	// and whether it will be an Array, Scalar, or either.
+	Descr() ValueDescr
+	// Type returns the datatype this expression will evaluate to.
+	Type() arrow.DataType
+
+	Hash() uint64
+	Equals(Expression) bool
+
+	// Bind binds this expression to the given input schema, looking up appropriate
+	// underlying implementations and some expression simplification may be performed
+	// along with implicit casts being inserted.
+	// Any state necessary for execution will be initialized.
+	//
+	// This only works in conjunction with cgo and being able to link against the
+	// C++ libarrow.so compute library. If this was not built with the libarrow compute
+	// support, this will panic.
+	Bind(context.Context, memory.Allocator, *arrow.Schema) (Expression, error)
+
+	// Release releases the underlying bound C++ memory that is allocated when
+	// a Bind is performed. Any bound expression should get released to ensure
+	// no memory leaks.
+	Release()
+
+	boundExpr() boundRef
+}
+
+func printDatum(datum Datum) string {
+	switch datum := datum.(type) {
+	case *ScalarDatum:
+		if !datum.Value.IsValid() {
+			return "null"
+		}
+
+		switch datum.Type().ID() {
+		case arrow.STRING, arrow.LARGE_STRING:
+			return strconv.Quote(datum.Value.(scalar.BinaryScalar).String())
+		case arrow.BINARY, arrow.FIXED_SIZE_BINARY, arrow.LARGE_BINARY:
+			return `"` + strings.ToUpper(hex.EncodeToString(datum.Value.(scalar.BinaryScalar).Data())) + `"`
+		}
+
+		return datum.Value.String()
+	default:
+		return datum.String()
+	}
+}
+
+// Literal is an expression denoting a literal Datum which could be any value
+// as a scalar, an array, or so on.
+type Literal struct {
+	Literal Datum
+
+	bound boundRef
+}
+
+func (Literal) FieldRef() *FieldRef     { return nil }
+func (l *Literal) String() string       { return printDatum(l.Literal) }
+func (l *Literal) boundExpr() boundRef  { return l.bound }
+func (l *Literal) Type() arrow.DataType { return l.Literal.(ArrayLikeDatum).Type() }
+func (l *Literal) IsBound() bool        { return l.Type() != nil }
+func (l *Literal) IsScalarExpr() bool   { return l.Literal.Kind() == KindScalar }
+
+func (l *Literal) Equals(other Expression) bool {
+	if rhs, ok := other.(*Literal); ok {
+		return l.Literal.Equals(rhs.Literal)
+	}
+	return false
+}
+
+func (l *Literal) IsNullLiteral() bool {
+	if ad, ok := l.Literal.(ArrayLikeDatum); ok {
+		return ad.NullN() == ad.Len()
+	}
+	return true
+}
+
+func (l *Literal) IsSatisfiable() bool {
+	if l.IsNullLiteral() {
+		return false
+	}
+
+	if sc, ok := l.Literal.(*ScalarDatum); ok && sc.Type().ID() == arrow.BOOL {
```

Comment:
so an error returns false?  What case does happen in?

Reply:
I'm not sure i understand your meaning. the logic here is:

```
if L is a null literal: return false
if L is a boolean scalar: return the value of the scalar bool
otherwise return true
```